rfcs/0005-nix-encryption.md

```diff
@@ -0,0 +1,273 @@
+  permission data. A NAR can no longer necessarily be unpacked in a
+  lossless way if the required uids/gids don't exist.
+
+* How should binary caches deal with ownership and permissions?
+
+Encryption support, on the other hand, is a very localised change,
+requiring only the addition of an encryptString primop and a
+decryption tool.
+
+## NixOps keys mechanism
+
+NixOps has a somewhat ad hoc mechanism for dealing with
+secrets. Secrets can be defined in the NixOps machine specification
+via the option `deployment.keys`:
+
+```
+deployment.keys.my-password.text = "fnord";
+```
+
+These secrets do not end up in the system closure; rather, they are
+copied by the command `nixops send-keys` to `/run/keys/<key-id>`.
+
+This mechanism could be generalised beyond NixOps into a Nix feature.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+## Reproducibility and nonces
+
+An important goal of NixOS is reproducibility. Nowadays this goal also
+includes bitwise-identical reproducibility (see
+https://reproducible-builds.org/); that is, building a NixOS system
+twice from the same Nix expressions should yield an identical result.
+
+However, libsodium's authenticated encryption requires supplying a
+nonce which (by definition) should be used only once. From
+https://download.libsodium.org/doc/secret-key_cryptography/authenticated_encryption.html:
+"The nonce doesn't have to be confidential, but it should never ever
+be reused with the same key."
+
+Unfortunately, the uniqueness of nonces is at odds with the
+requirement of bitwise reproducibility: if multiple calls to
+`encryptString` with the same inputs return different results, then
+logically identical NixOS system closures will not be bitwise
+identical.
+
+We could replace the nonce with a deterministic value, but it's not
+entirely clear what the cryptographic implications are. At the very
+least, it allows attackers to obverse that a secret has changed, or
+that it has changed back to a previously observed value.
+
+## Decryption failure
+
+What should `nix decrypt` do if it cannot decrypt an encrypted string?
+Keeping the encrypted string or replacing it by an empty string is
+dangerous because it can cause e.g. a predictable password to be
+emitted. Probably the best solution is to fail by default (i.e. not
+emit a file at all), or optionally replace the encrypted string by a
+cryptographically secure random string.
+
+## Public-key cryptography
+
+We could have `encryptString` use a public key, and `nix decrypt` use
+the corresponding private key. This is kind of nice because Nix
+already has a command to generate public/private key pairs using
+libsodium for binary cache signing.
+
+However, I couldn't think of a plausible use case for this. It would
+allow the user who generates/builds the configuration not to have the
+decryption key, but that doesn't seem very useful.
+
+## Key generation
+
+Currently, we assume that `<nixos-store-key>` exists beforehand. It
```

Comment:
How would this work in a multi-user environment? Could each user feasibly have their own key? And how would this integrate with something like NixOps, where one might want to have a separate key for NixOps deployments, and possibly share that key with others managing the deployment, without exposing their local key?